File: core/java/src/main/java/hara/lang/protocol/ICas.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Atomically replaces an expected value. */
@HaraProtocolBinding(namespace = "std.protocol.icas", name = "ICas")
public interface ICas<V> {
  @HaraMethod(value = "cas", arity = 3)
  boolean cas(V oldValue, V newValue);
}

File: core/rust/src/lang/protocol/icas.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.icas", name = "ICas")]
pub trait ICas<V> {
    type Error;

    #[hara_method(value = "cas", arity = 3)]
    fn cas(&self, old_value: &V, new_value: V) -> Result<bool, Self::Error>;
}
